File: Departamentos/Empregados.hs
data Empregados = Empregados {
    nome :: String,
    id :: Int,
    cargo :: String,
    salario :: Float,
    departamento :: Maybe Departamento
} deriving (Show)

data Departamento = Departamento {
    nomeD :: String,
    idD :: Int,
    localizacao :: String,
    empregados :: [Empregados]
} deriving (Show)

atualizarDados :: Empregados -> String -> String -> Float -> Empregados
atualizarDados emp novoNome novoCargo novoSalario = emp { nome = novoNome, cargo = novoCargo, salario = novoSalario }

transferencia :: Departamento -> Empregados -> Empregados
transferencia depto emp = emp { departamento = Just depto }

imprimirDados :: Empregados -> IO ()
imprimirDados emp = do
    putStrLn $ "Nome: " ++ nome emp
    putStrLn $ "ID: " ++ show (id emp)
    putStrLn $ "Cargo: " ++ cargo emp
    putStrLn $ "Salário: " ++ show (salario emp) ++ "\n"

imprimirEmpregados :: Departamento -> IO ()
imprimirEmpregados depto = do
    putStrLn "Lista de Empregados:\n"
    mapM_ imprimirDados (empregados depto)

informacoesDepartamento :: Departamento -> IO ()
informacoesDepartamento depto = do
    putStrLn $ "Nome: " ++ nomeD depto
    putStrLn $ "ID: " ++ show (idD depto)
    putStrLn $ "Localizacao: " ++ localizacao depto

main :: IO ()
main = do
    let alberto = Empregados "Alberto" 18 "Atendente" 1300.00 Nothing
        carlos = Empregados "Carlos" 18 "Atendente" 1300.00 Nothing
        recepcao = Departamento "Recepção" 1 "Entrada" [alberto, carlos]

    let albertoTransferido = transferencia recepcao alberto
        carlosTransferido = transferencia recepcao carlos

    putStrLn "Empregados na Recepção:"
    imprimirEmpregados recepcao

    putStrLn "\nInformações do Departamento:"
    informacoesDepartamento recepcao

File: Departamentos/Empregados.java
import java.util.ArrayList;

public class Empregados {
    String nome;
    int id;
    String cargo;
    float salario;
    Departamento departamento;

    public Empregados(String nome, int id, String cargo, float salario) {
        this.nome = nome;
        this.id = id;
        this.cargo = cargo;
        this.salario = salario;
    }

    public void atualizarDados(String nome, String cargo, float salario) {
        this.nome = nome;
        this.cargo = cargo;
        this.salario = salario;
    }
    
    public void transferencia(Departamento departamento, Empregados empregado){
        this.departamento = departamento;
        departamento.adicionarEmpregado(empregado);
    }
    
    public void imprimirDados(){
        System.out.println("Nome: " + this.nome);
        System.out.println("ID: " + this.id);
        System.out.println("Cargo: " + this.cargo);
        System.out.println("Salário: " + this.salario);
        System.out.println("");
    }
    
    public void imprimirDepartamento(){
        departamento.informacoesDepartamento();
    }
}

public class Departamento {
    String nome;
    int id;
    String localizacao;
    ArrayList<Empregados> empregados;
    
    public Departamento(String nome, int id, String localizacao){
        this.nome = nome;
        this.id = id;
        this.localizacao = localizacao;
        this.empregados = new ArrayList<>();
    }
    
    public void adicionarEmpregado(Empregados empregado){
        this.empregados.add(empregado);
    }
    
    public void imprimirEmpregados(){
        System.out.printf("Lista de Empregados:\n\n");
        for(Empregados pessoa : empregados){
            pessoa.imprimirDados();
        }
    }
    
    public void removerEmpregado(Empregados empregado){
        this.empregados.remove(empregado);
    }
    
    public void informacoesDepartamento(){
        System.out.println("Nome: " + this.nome);
        System.out.println("ID: " + this.id);
        System.out.println("Localização: " + this.localizacao);
    }
}

public class App {
    public static void main(String[] args) throws Exception {
        Empregados Alberto = new Empregados("Alberto", 18, "Atendente", 1300.00f);
        Empregados Carlos = new Empregados("Carlos", 18, "Atendente", 1300.00f);
        Departamento Recepcao = new Departamento("Recepção", 1, "Entrada");
        Carlos.transferencia(Recepcao, Carlos);
        Alberto.transferencia(Recepcao, Alberto);
        Recepcao.imprimirEmpregados();
        Alberto.imprimirDepartamento();
    }
}
